### DISCUSSION TITLE
write_gml, preserve desired node id

### DISCUSSION THREAD

**Discussion by mattall** [networkx/networkx#5812] [Ideas]
I am frequently reading/writing GML files for projects, and I have a map of node ids that I'd like to be consistent when i write a graph after making some minor changes to it. 

After I have labeled my nodes by hand, networkx.write_gml re-indexes them for me, and I'm left with a gml file with items shown below.
There is a 1-to-1 mapping from ids to labels, and they have the same range. 
I've gone through myself and written my own write_gml function to preserve my indexing but it's not as versatile or efficient as networkx.write_gml. 
I think it would be a nice feature if networkx.write_gml had an optional flag to refrain from re-indexing the graph. 

  node [
    id 0
    label "89"
    Latitude 37.3361905
    Longitude -121.8905833
  ]
  node [
    id 1
    label "83"
    Latitude 36.6744117
    Longitude -121.6550372
  ]
...
  node [
    id 29
    label "29"
    Latitude 44.0505054
    Longitude -123.0950506
  ]

**Comment by dschult** [comment]
We do provide that feature when reading a GML graph. That is, we relabel the nodes to be the value of the "label" attribute of those nodes, and we allow the user to choose which attribute is the "label".

Are you needing the `id`s to match your mapping for some other program to manipulate them? Or is it just for storage and reading back into NetworkX?

**Comment by mattall** [comment]
Yes, I need to keep my mapping consistent for debugging. It just makes it easier to verify that operations such as {add/remove}_edge are getting applied to the objects that I'm intending them to.

I've been just using this script (with minor modifications per task or project) to keep my index and labeling consistent 

    def write_gml(G, name):
      with open(name, 'w') as fob:
        fob.write('graph [\n')
        for node in sorted(G.nodes()):
            str_node = str(node)
            id = str_node.strip('s')
            fob.write('\tnode [\n')
            fob.write('\t\tid {}\n'.format(id))
            fob.write('\t\tlabel "{}"\n'.format(node))
            for key, value in G.nodes[node].items():
                fob.write('\t\t{} {}\n'.format(key, value))
            fob.write('\t]\n')
        for s, t in G.edges():
            str_s, str_t = str(s), str(t)
            src = str_s.strip('s')
            dst = str_t.strip('s')
            fob.write('\tedge [\n')
            fob.write('\t\tsource {}\n'.format(src))
            fob.write('\t\ttarget {}\n'.format(dst))
            for key, value in G.edges[s,t].items():
                fob.write('\t\t{} {}\n'.format(key, value))
            fob.write('\t]\n')
        fob.write(']')
        return
    



`

**Comment by rossbar** [comment]
IIUC the problem is that information that you'd hoped would be immutable is being stored in the `id` and/or `label` node attributes. Note that these attributes have special meaning in the GML format, as is noted in the `write_gml` docstring:

> Notes
> -----
> Graph attributes named 'directed', 'multigraph', 'node' or
> 'edge', node attributes named 'id' or 'label', edge attributes
> named 'source' or 'target' (or 'key' if `G` is a multigraph)
> are ignored because these attribute names are used to encode the graph
> structure.

In other words, the *specific* node attributes `"id"` and `"label"` have special meaning in the GML format, so user-information should not be stored on them. The easiest way to guarantee the node data will round-trip correctly is to give it a different attribute name that doesn't conflict with the special gml keywords; e.g. `"nid"` and `"nlabel"`.

AIUI these "special" attribute names come from GML itself, so overriding this behavior might actually be in violation of the standard and result in invalid gml data files.
